### PR TITLE
Add power mode control and reaction

### PR DESCRIPTION
**Pull request by brunoais**
Adds the ability to change the power mode of the PC through various means and also sends power mode updates to interested parties through a character device.

### Rundown:

1. React to presses to the power mode button and change it.
2. Add sysfs files to read and write power mode (and also what options are available).
3. Allow changing power mode through ioctl (`W_UW_POWER_MODE`)
4. Allow listening to power mode changes so the PC can react to them.
    1. This device is also expected to be used for reporting other events such as a max-fan speed button (if it exists)

### Power mode in manual hardware mode does nothing

From what I demonstrated on video in #86, if the hardware is set to manual mode, which tuxedo-keyboard driver does by default,[ the power mode setting that affects the LEDs doesn't do anything.](https://github.com/tuxedocomputers/tuxedo-keyboard/issues/86#issuecomment-839510602)
I was also able to confirm that by using [`ryzenadj`](https://github.com/FlyGoat/RyzenAdj) with the `-i` command line argument.
However, if I turn off the manual mode, I do see the changes in the values.

### Help needed:
I only own a single laptop, the Polaris 17''. I need help knowing how other laptops with other power modes work. The code is ready to have them setup but I can't test and do it myself because I only own a single sample.
@maintainers: Please add a commit with the changes required for other laptops.

### Relicensing
I accept relicensing of this code to GPLv2 when that comes around.

I automatically accept relicensing to a different open-source license if I'm prompted 3 times over a period of 1 year and I don't answer. `@` me on an issue in this repository 3 times in 3 different messages is enough.

### CODE REVIEW

File: src/uniwill_keyboard.h [hunk 1]
```diff
@@ -57,6 +57,25 @@
...
 #define UNIWILL_COLOR_DEFAULT			0xffffff
 
+
+#define UNIWILL_POWER_MODE_MASK 0x90
+#define UNIWILL_POWER_MODE_MIN 0x80
+#define UNIWILL_POWER_MODE_MED 0x00
+#define UNIWILL_POWER_MODE_MAX 0x90
+#define UNIWILL_POWER_MODE_MAX_ALT 0x10
+#define UNIWILL_POWER_MODE_UNSET(var) (var) & ( ~ UNIWILL_POWER_MODE_MASK)
+#define UNIWILL_POWER_MODE(var) (var) & UNIWILL_POWER_MODE_MASK
```
Comment (brunoais) [inline src/uniwill_keyboard.h:67]: These are the power modes I know

File: src/uniwill_keyboard.h [hunk 2]
```diff
@@ -504,9 +617,109 @@ static int uw_kbd_bl_init(struct platform_device *dev)
...
+		// | dmi_match(DMI_BOARD_NAME, "Polaris1701I2060")
+		;
+
+
+	if (!uniwill_has_power_mode) {
+		TUXEDO_INFO("PC without power mode");
+		return 127;
 	}
 
+	// Init sysfs power mode attributes group
```
Comment (brunoais) [inline src/uniwill_keyboard.h:718]: Power modes need to be written before this comment.
Setting the different values depending on the laptop and the power modes it allows.
For that, edit `power_modes` and `power_modes_count` at these lines.
Those variables are defined near the top of the file at: https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105/files#diff-e9ab76a5901bf8345c65cfbdd9fc9d43663ccb17544b1dcef52d141b9b3630c6R70-R75



You can find the addresses I know here:
https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105/files#diff-e9ab76a5901bf8345c65cfbdd9fc9d43663ccb17544b1dcef52d141b9b3630c6R61-R67

Comment (brunoais) [inline src/uniwill_keyboard.h:718]: Power modes need to be written before this comment.
Setting the different values depending on the laptop and the power modes it allows.
For that, edit `power_modes` and `power_modes_count` at these lines.
Those variables are defined near the top of the file at: https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105/files#diff-e9ab76a5901bf8345c65cfbdd9fc9d43663ccb17544b1dcef52d141b9b3630c6R70-R75



You can find the addresses I know here:
https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105/files#diff-e9ab76a5901bf8345c65cfbdd9fc9d43663ccb17544b1dcef52d141b9b3630c6R61-R67

File: src/uniwill_keyboard.h (hunk 1 above)
Comment (brunoais) [inline src/uniwill_keyboard.h:67]: These are the power modes I know

File: src/uniwill_keyboard.h [hunk 3]
```diff
@@ -57,6 +57,25 @@
...
+#define UNIWILL_POWER_MODE_UNSET(var) (var) & ( ~ UNIWILL_POWER_MODE_MASK)
+#define UNIWILL_POWER_MODE(var) (var) & UNIWILL_POWER_MODE_MASK
+
+
+static u8 power_modes[] = {
+	UNIWILL_POWER_MODE_MIN, 
+	UNIWILL_POWER_MODE_MED, 
+	UNIWILL_POWER_MODE_MAX,
+};
+static u8 power_modes_count = sizeof(power_modes)/sizeof(power_modes[0]);
```
Comment (brunoais) [inline src/uniwill_keyboard.h:75]: These are the variables to edit in the code

**Review by brunoais — COMMENTED** [review]
Help is needed here from @tuxedoxt

**Comment by brunoais** [comment]
@tuxedoxt I fixed the code with the new update. Feel free to share with me anything that I missed.

I can't make adjustments for the needs of laptops I don't own, so feel free to submit commits that take them into account

**Comment by brunoais** [comment]
@tuxedoxt Any updates?

**Comment by tuxedoxt** [comment]
> @tuxedoxt Any updates?

Hello,

sorry, this PR seems to have been overlooked. You're right that it's hard to design functionality based on only one device.

Alright, there's a lot happening here.

Regarding the original point (1). I believe we want to use this to change TCC profiles in the future, which in turn can have "odm profiles" assigned. However that needs to go through userspace in this case. Maybe your version could be used in a "non manual mode" scenario.

The *io_announce* functionality is interesting, what do you use this for? It might be possible to trigger this from an event instead of by manual set.

Since the previous discussions on(/in) the issue our own investigations moved on as well. It turns out there are a few cases that need to be addressed. Here's an update.

- Two profiles (e.g Pulse, BA15) no LEDs
- Three profiles (e.g Polaris gen 1) with LEDs

These devices were thought to couple LEDs (where available) and profiles and to some extent fan control (without "full fan mode"). Further testing seems to show however that at least some have the properties of the following categories (power control separate).

After this the new devices arrived and I had a look at 

- Two profiles (IBP14gen6) no LEDs
- Three profiles (Polaris/Stellaris gen 2/3) with LEDs

However, for these the profiles do not cover power limits directly. The limits PL1, PL2, (PL4) can be set individually regardless of "odm profile". For these devices LEDs are actually controllable independently, at least from power limits.

It's quite the mess. This led to a basic implementation (driver + TCC) for the currently known cases, that is on the way to release. However, there are many devices to test.

**Comment by brunoais** [comment]
> > @tuxedoxt Any updates?
> 
> Hello,
> 
> sorry, this PR seems to have been overlooked. You're right that it's hard to design functionality based on only one device.

Apologies accepted.

> Regarding the original point (1). I believe we want to use this to change TCC profiles in the future, which in turn can have "odm profiles" assigned. However that needs to go through userspace in this case. Maybe your version could be used in a "non manual mode" scenario.

Could be. However, I like being able to control it from userspace. I do that for my "night power mode"  script. It confirms night mode by swapping between 1 and 0 for a short time (effectively, blinking 1 LED) and then leaving on 0. This makes it the slowest performing for maintenance reasons (most efficient processing when responsiveness isn't needed).
This is just one of my uses.

> The _io_announce_ functionality is interesting, what do you use this for? It might be possible to trigger this from an event instead of by manual set.

I use that to announce to a python daemon (see: https://github.com/brunoais/Tuxedo-Keyboard-power-mode-daemon). In my use, it calls `ryzenadj` with predefined values to change the PC's performance (and energy-efficiency; not time efficiency).
I also made time-sensitive scripts that cause different power reactions based on time of day. E.g. "night mode" is only available since 9pm. Before that, "night mode" is not considered.

When you mention "from an event", do you mean pressing the pwoer button? It already works with this patch. See: https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105/files#diff-e9ab76a5901bf8345c65cfbdd9fc9d43663ccb17544b1dcef52d141b9b3630c6R435

> Since the previous discussions on(/in) the issue our own investigations moved on as well. It turns out there are a few cases that need to be addressed. Here's an update.
> 
>     * Two profiles (e.g Pulse, BA15) no LEDs
> 
>     * Three profiles (e.g Polaris gen 1) with LEDs

The code is ready for that. We may need to hard code how many modes each has in the driver itself (pretty much the norm around now). However, the code is compatible with a quick variable setting to define how many options are available (hoping that the addresses are the same). (see: https://github.com/tuxedocomputers/tuxedo-keyboard/pull/105#discussion_r706815039))

For the ones without LED, that information can be provided by a daemon which listens for events.
 
> These devices were thought to couple LEDs (where available) and profiles and to some extent fan control (without "full fan mode"). Further testing seems to show however that at least some have the properties of the following categories (power control separate).

Fan control and power control separate? I have no idea if the Polaris has those separate. I've tried to check but there doesn't seem to exist a function activatable (E.g. Fn + 1) to change the fan speeds. I've tried that with auto mode engaged (and tuxedo-control-center turned off).
For my laptop, ryzenadj allows for the best fine-tuning power management I ever dealt with. The options in tuxedo-control-center are very limiting in comparison (FYI: I do understand why it's that limited).

> After this the new devices arrived and I had a look at
> 
>     * Two profiles (IBP14gen6) no LEDs
> 
>     * Three profiles (Polaris/Stellaris gen 2/3) with LEDs
> 
>
> However, for these the profiles do not cover power limits directly. The limits PL1, PL2, (PL4) can be set individually regardless of "odm profile". For these devices LEDs are actually controllable independently, at least from power limits.
> 
> It's quite the mess. This led to a basic implementation (driver + TCC) for the currently known cases, that is on the way to release. However, there are many devices to test.

The most important parts as of now is understanding if the addresses are the same. If they are, the solution is easier. If they are different, then some work can be done to accommodate that.

Because this change is only for uniwill, hopefully the addresses are the same. It should also make it easier for them to make the windows drivers. Maybe even the bits that represent states in some laptops that other laptops don't have are just ignored.

I'm available to help in whatever I can. However, I am very much interested in this kind of feature to be included.

**Comment by brunoais** [comment]
@tuxedoxt Any updates?

**Comment by tuxedoxt** [comment]
> @tuxedoxt Any updates?

Hey, yes,the power control features got even more delayed but are now finally released as of TCC v1.2.0 + TK v3.1.0.

For Polaris Gen 1 there are profiles connected to the mode LED and the led moves with selected profile.

For Polaris/Stellaris Gen 2/3/4 there are TDP limit control for power and the profiles only switch the LED. For these devices we have allowed setting the profile as well but visually selecting LED state instead.

For a list of features and devices see change log

https://github.com/tuxedocomputers/tuxedo-control-center/blob/master/CHANGELOG.md#120---2022-10-10

Thanks for the feedback on the topic so far. Even if it's hard to collaborate code-wise on this (PR-wise) the feedback is very valuable!

Please let me know how it works for your use case, if you have thoughts on the feature and how to proceed from here.

**Comment by brunoais** [comment]
> Please let me know how it works for your use case, if you have thoughts on the feature and how to proceed from here.

I still want to think more throughfully because your design is way different than my design.

@tuxedoxt But I rushed to write something already because there's a crash-worthy or breakage-worthy bug in the code (not applicable for my laptop because it has the 3 power modes, so I can't know for sure):
https://github.com/tuxedocomputers/tuxedo-keyboard/issues/140

**Comment by brunoais** [comment]
@tuxedoxt For example, something I'd do differently:

Instead of:

	struct uniwill_device_features_t {
		u8 model;
		/**
		 * Identification for uniwill_power_profile_v1
		 *
		 * - Two profiles present in low power devices often called
		 *   "power save" and "balanced".
		 * - Three profiles present mainly in devices with discrete
		 *   graphics card often called "power save", "balanced"
		 *   and "enthusiast"
		 */
		bool uniwill_profile_v1;
		bool uniwill_profile_v1_two_profs;
		bool uniwill_profile_v1_three_profs;
		bool uniwill_profile_v1_three_profs_leds_only;
	};

I'd do:

	struct uniwill_device_features_t {
		u8 model;
		/**
		 * Identification for uniwill_power_profile_v1
		 *
		 * - Two profiles present in low power devices often called
		 *   "power save" and "balanced".
		 * - Three profiles present mainly in devices with discrete
		 *   graphics card often called "power save", "balanced"
		 *   and "enthusiast"
		 */
		u8 num_leds;
		u8 (*uniwill_get_profile)();
		bool (*uniwill_set_profile)(u8);
		bool leds_only;
	};


Then `uniwill_get_device_features()` would set the correct functions as appropriate.
This code would be much more ready for future situations with different (or even more) power profiles or whatever inventions are made additionally.

FYI: I'm still analysing the code. That's just something that kept "ringing2 in my head

**Comment by brunoais** [comment]
@tuxedoxt I'm still waiting for feedback on your plans on exposing this hardware to other programs.
I'm also waiting to know if you will provide more API means to expose waiting for power mode changes (LED) besides a wasteful polling method instead of event/blocking wait.

**Pull request closed without merging**